Fix CUDA stream syncing bug in allgather and reduce_scatter

mxw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#19631 Fix CUDA stream syncing bug in allgather and reduce_scatter**

Differential Revision: [D15110077](https://our.internmc.facebook.com/intern/diff/D15110077)

mxw committed:
Fix CUDA stream syncing bug in allgather and reduce_scatter

mrshenli commented:
Thanks for fixing this!! Could you please remove irrelevant changes to the `third_party` folder? I usually do the following:

```
git clean -xdf
git submodule deinit -f .
git submodule sync
git submodule update --init --recursive
```

mrshenli reviewed (approved):
approve to unblock, but please do clean `third_party` folder.

mxw committed:
Update on "Fix CUDA stream syncing bug in allgather and reduce_scatter"

Fix CUDA stream syncing bug in allgather and reduce_scatter

gh-metadata: pytorch pytorch 19631 gh/mxw/4/head

mxw committed:
Update on "Fix CUDA stream syncing bug in allgather and reduce_scatter"

Fix CUDA stream syncing bug in allgather and reduce_scatter

gh-metadata: pytorch pytorch 19631 gh/mxw/4/head

mxw committed:
Update on "Fix CUDA stream syncing bug in allgather and reduce_scatter"

Fix CUDA stream syncing bug in allgather and reduce_scatter

gh-metadata: pytorch pytorch 19631 gh/mxw/4/head

mxw commented:
@pytorchbot retest this please

mxw commented:
@pytorchbot retest this please
